server/oidc.go

```diff
@@ -0,0 +1,269 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Genome Research Ltd.
+ *
+ * Author: Michael Grace <mg38@sanger.ac.uk>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ ******************************************************************************/
+
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+	"github.com/gorilla/sessions"
+	verifier "github.com/okta/okta-jwt-verifier-golang"
+	oauthUtils "github.com/okta/okta-jwt-verifier-golang/utils"
+	"github.com/thanhpk/randstr"
+
+	"golang.org/x/oauth2"
+)
+
+const oktaCookieName = "okta-hosted-login-session-store"
+
+var sessionStore = sessions.NewCookieStore([]byte(oktaCookieName))
+
+type oAuthParameters struct {
+	issuer       string
+	clientID     string
+	clientSecret string
+}
+
+// newOktaOauthConfig returns a oauth2.Config with the appropriate values based
+// on defaults, the paramenters in oAuthParameters and the callack URL
+func (p oAuthParameters) newOktaOauthConfig(callback string) oauth2.Config {
```

Comment:
"toOkta" feels like a better name for this. We're sort of converting one config struct to another.

Docs end in a full stop. Should run `make lint`.